src/connection.cpp

```diff
@@ -350,120 +395,172 @@ namespace oxen::quic
+                // ndatalen "should" be >0 for streams and <0 for psuedo streams
+                if (nwrite > 0 && ((stream_id == -1) ? ndatalen < 0 : ndatalen >= 0))
                 {
-                    log::trace(log_cat, "Consumed {} bytes from stream {} and have space left", ndatalen, stream_id);
-                    assert(ndatalen >= 0);
-                    if (stream)
-                        stream->wrote(ndatalen);
-                    // If we had more data on the stream, we wouldn't have got a WRITE_MORE, so
-                    // don't need to re-add the stream to strs.
+                    write_successful = true;
+                    log::trace(log_cat, "consumed {} bytes from stream {}", ndatalen, stream_id);
+                    if (stream_id != -1)
+                        source->wrote(ndatalen);
                 }
-                else if (nwrite == NGTCP2_ERR_CLOSING)  // -230
-                    log::debug(log_cat, "Cannot write to {}: connection is closing", stream_id);
-                else if (nwrite == NGTCP2_ERR_STREAM_SHUT_WR)  // -221
-                    log::debug(log_cat, "Cannot add to stream {}: stream is shut, proceeding", stream_id);
-                else if (nwrite == NGTCP2_ERR_STREAM_DATA_BLOCKED)  // -210
-                    log::debug(log_cat, "Cannot add to stream {}: stream is blocked", stream_id);
-                else
-                    log::error(log_cat, "Error writing stream data: {}", ngtcp2_strerror(nwrite));
-
-                continue;
             }
-
-            if (nwrite == 0)  // we are congested (or done)
+            else  // datagram block
             {
-                log::trace(
-                        log_cat,
-                        "Done stream writing to {} ({}connection is congested)",
-                        stream_id,
-                        stream ? "" : "nothing else to write or ");
-                if (stream)
-                    // we are congested, so clear all pending streams (aside from the -1
-                    // pseudo-stream at the end) so that our next call hits the -1 to finish off.
-                    strs.erase(strs.begin(), streams_end_it);
-                continue;
+                nwrite = ngtcp2_conn_writev_datagram(
+                        conn.get(),
+                        _path,
+                        &pkt_info,
+                        buf_pos,
+                        max_datagram_size,
+                        &datagram_accepted,
+                        flags,
+                        0,
+                        bufs.data(),
+                        1,
+                        ts);
+
+                log::debug(log_cat, "ngtcp2_conn_writev_datagram returned a value of {}", nwrite);
+
+                // both must be true to be successful
+                if (datagram_accepted != 0)
+                {
+                    // if the datagram was accepted at all, drop it from the buffer
```

Comment:
```suggestion
                    // if the datagram was accepted, drop it from the buffer
```
The "at all" implies it could be partially accepted, but it cannot be.